Add association setup message

w910820618 opened the pull request:
This time I encapsulated the association setup message. Because I won't use your test tool, I have updated the test data package synchronously, you can check it.

w910820618 commented:
```golang
package main

import (
	"flag"
	"github.com/wmnsk/go-pfcp/ie"
	"github.com/wmnsk/go-pfcp/message"
	"log"
	"net"
	"time"
)

func main() {
	var (
		server = flag.String("-s", "127.0.0.2:8805", "server's addr/port")
	)
	flag.Parse()

	raddr, err := net.ResolveUDPAddr("udp", *server)
	if err != nil {
		log.Fatal(err)
	}

	conn, err := net.DialUDP("udp", nil, raddr)
	if err != nil {
		log.Fatal(err)
	}
	nid := ie.NewNodeID("172.55.55.101", "", "")
	fseid := ie.NewFSEID(0xbebbebebebbebebf, net.IPv4(172, 55, 55, 101), nil, nil)

	pdrid := ie.NewPacketDetectionRuleID(1)
	precedence := ie.NewPrecedence(0)

	sourceInterface := ie.NewSourceInterface(ie.SrcInterfaceAccess)
	fteid := ie.NewFTEID(0, nil, nil, nil)
	ueIPAddress := ie.NewUEIPAddress(0x02, "10.0.0.3", "", 0)
	pdi := ie.NewPDI(0, sourceInterface, fteid, ueIPAddress)
	outerHeaderRemoval := ie.NewOuterHeaderRemoval(0, 0)
	farid := ie.NewFARID(1)
	createPDR := ie.NewCreatePDR(0, pdrid, precedence, pdi, outerHeaderRemoval, farid)

	applyAction := ie.NewApplyAction(0x02)
	destinationInterface := ie.NewDestinationInterface(ie.DstInterfaceCore)
	outerHeaderCreation := ie.NewOuterHeaderCreation(0x0100, 0x00000001, "192.168.20.136", "", 0, 0, 0)
	forwardingParameters := ie.NewForwardingParameters(0, destinationInterface, outerHeaderCreation)
	createFAR := ie.NewCreateFAR(0, farid, applyAction, forwardingParameters)
	hbreq, err := message.NewSessionEstablishmentRequset(nid, fseid, createPDR, createFAR).Marshal()
	if err != nil {
		log.Fatal(err)
	}

	if _, err := conn.Write(hbreq); err != nil {
		log.Fatal(err)
	}
	log.Printf("sent Heartbeat Request to: %s", raddr)

	if err := conn.SetReadDeadline(time.Now().Add(3 * time.Second)); err != nil {
		log.Fatal(err)
	}
}

func NewCreateFAR(teid uint16, farID *IE, applyAction *IE, forwardingParameter *IE) *IE {
	return newGroupedIE(CreateFAR, teid, farID, applyAction, forwardingParameter)
}

func NewCreatePDR(teid uint16, pdrid *IE, precedence *IE, pdi *IE, outerHeaderRemoval *IE, farID *IE) *IE {
	return newGroupedIE(CreatePDR, teid, pdrid, precedence, pdi, outerHeaderRemoval, farID)
}

func NewForwardingParameters(teid uint16, destinationInterface *IE, outerHeaderCreation *IE) *IE {
	return newGroupedIE(ForwardingParameters, teid, destinationInterface, outerHeaderCreation)
}

func NewPDI(teid uint16, sourceInterface *IE, fteid *IE, ueIpAddress *IE) *IE {
	return newGroupedIE(PDI, teid, sourceInterface, fteid, ueIpAddress)
}
```

[pfcp.pcap.zip](https://github.com/wmnsk/go-pfcp/files/4415449/pfcp.pcap.zip)

wmnsk commented on examples/associationsetup/ac-server/main.go:
```diff
@@ -0,0 +1,72 @@
+			continue
+		}
+
+		ts, err := acreq.NodeID.NodeID()
+		if err != nil {
+			log.Printf("got Association Setup Request with invalid TS: %s, from: %s", err, addr)
+			continue
+		} else {
+			log.Printf("got Association Setup Request with TS: %s, from: %s", ts, addr)
```
`ts` and `TS` meant "timestamp" in heartbeat example. How about using `n` and `NodeID` instead?

```suggestion
		n, err := acreq.NodeID.NodeID()
		if err != nil {
			log.Printf("got Association Setup Request with invalid NodeID: %s, from: %s", err, addr)
			continue
		} else {
			log.Printf("got Association Setup Request with NodeID: %s, from: %s", n, addr)
```

w910820618 replied:
ok!

wmnsk commented on examples/associationsetup/ac-client/main.go:
```diff
@@ -0,0 +1,75 @@
+		}
+
+		waiting = false
+		ts, err := acres.NodeID.NodeID()
+		if err != nil {
+			log.Printf("got Association Setup  Response with invalid TS: %s, from: %s", err, addr)
+			break
+		}
+		log.Printf("got Association Setup  Response with TS: %s, from: %s", ts, addr)
```
Same as the suggestion in server side.

```suggestion
		n, err := acres.NodeID.NodeID()
		if err != nil {
			log.Printf("got Association Setup  Response with invalid NodeID: %s, from: %s", err, addr)
			break
		}
		log.Printf("got Association Setup  Response with NodeID: %s, from: %s", n, addr)
```

wmnsk commented on examples/associationsetup/ac-client/main.go:
```diff
@@ -0,0 +1,78 @@
+// Copyright 2019-2020 go-pfcp authors. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+package main
```
You need one more empty line so that the Copyright notice is not interpreted as GoDoc(same for other files).

Also, could you add brief description of this command like [this](https://github.com/wmnsk/go-pfcp/blob/master/examples/heartbeat/hb-client/main.go#L5-L8)?

wmnsk commented on examples/associationsetup/ac-server/main.go:
```diff
@@ -0,0 +1,76 @@
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+
+package main
```
Same here, please add brief description of this command like [this](https://github.com/wmnsk/go-pfcp/blob/master/examples/heartbeat/hb-client/main.go#L5-L8).

wmnsk commented on message/assocaition-setup-request_test.go:
```diff
@@ -0,0 +1,4 @@
+// Copyright 2019-2020 go-pfcp authors. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+package message_test
```
Empty line

wmnsk commented on message/association-setup-request.go:
```diff
@@ -0,0 +1,147 @@
+// Copyright 2019-2020 go-pfcp authors. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+package message
```
Empty line

wmnsk commented on message/association-setup-request.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// NewAssociationSetupRequest creates a new AssociationSetupRequest.
+func NewAssociationSetupRequest(nid *ie.IE, IEs ...*ie.IE) *AssociationSetupRequest {
```
Don't capitalize the name of parameter.

```suggestion
func NewAssociationSetupRequest(nid *ie.IE, ies ...*ie.IE) *AssociationSetupRequest {
```

wmnsk commented on examples/associationsetup/ac-client/main.go:
```diff
@@ -0,0 +1,78 @@
+			log.Printf("got Association Setup  Response with invalid NodeID: %s, from: %s", err, addr)
+			break
+		}
+		log.Printf("got Association Setup  Response with NodeID: %s, from: %s", nid, addr)
```
Excessive whitespace.

```suggestion
		log.Printf("got Association Setup Response with NodeID: %s, from: %s", nid, addr)
```

wmnsk commented on examples/associationsetup/ac-client/main.go:
```diff
@@ -0,0 +1,78 @@
+	if _, err := conn.Write(acreq); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("sent AssociationSetup  Request to: %s", raddr)
```
Excessive whitespace.

```suggestion
	log.Printf("sent AssociationSetup Request to: %s", raddr)
```

wmnsk commented on message/association-setup-request.go:
```diff
@@ -0,0 +1,147 @@
+
+// MessageTypeName returns the name of protocol.
+func (m *AssociationSetupRequest) MessageTypeName() string {
+	return "PFCP Association Setup Request"
```
```suggestion
	return "Association Setup Request"
```

wmnsk commented on message/association-setup-response.go:
```diff
@@ -0,0 +1,144 @@
+
+// MessageTypeName returns the name of protocol.
+func (m *AssociationSetupResponse) MessageTypeName() string {
+	return "PFCP Association Setup Response"
```
```suggestion
	return "Association Setup Response"
```

wmnsk commented on message/association-setup-response_test.go:
```diff
@@ -0,0 +1,4 @@
+// Copyright 2019-2020 go-pfcp authors. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+package message_test
```
Empty line

w910820618 commented:
I have modified some of the issues you mentioned, and I changed the packaging logic of outerHeaderCreation. I refer to the code of wireshark. There are some issues with the unit tests, and I'm still fixing them. [wireshark pfcp](https://fossies.org/linux/wireshark/epan/dissectors/packet-pfcp.c)
